### PR TITLE
refactor: parquet pruning simplifications

### PR DESCRIPTION
**Pull request by crepererum**
# Which issue does this PR close?
Prep work for #4695.

# Rationale for this change
Makes the actual change in #4695 easier.

# What changes are included in this PR?
Internal improvements.

# Are these changes tested?
Existing tests pass, no function changes.

# Are there any user-facing changes?
\-

### CODE REVIEW

**Comment by crepererum** [comment]
I am pretty sure that this test breakage is unrelated, seems flaky to me.

File: datafusion/core/src/physical_plan/file_format/parquet/page_filter.rs [hunk 1]
```diff
@@ -110,14 +110,16 @@ impl PagePruningPredicate {
...
-            .filter_map(|predicate| match predicate.to_columns() {
-                Ok(columns) if columns.len() == 1 => {
-                    match PruningPredicate::try_new(predicate.clone(), schema.clone()) {
-                        Ok(p) if !p.allways_true() => Some(Ok(p)),
-                        _ => None,
+            .filter_map(|predicate| {
+                match PruningPredicate::try_new(predicate.clone(), schema.clone()) {
+                    Ok(p)
+                        if (!p.allways_true())
+                            && (p.required_columns().n_columns() < 2) =>
```
Comment (avantgardnerio) [inline datafusion/core/src/physical_plan/file_format/parquet/page_filter.rs:117]: This is a behavior change for `n_columns() == 0`. Based on:

```
    pub fn allways_true(&self) -> bool {
        self.predicate_expr
            .as_any()
            .downcast_ref::<Literal>()
            .map(|l| matches!(l.value(), ScalarValue::Boolean(Some(true))))
            .unwrap_or_default()
    }
```

I ran the test suite, panicing if `n_columns() == 0` and I can't get it to happen, so I guess it LGTM.
I assume that would default to false, in which case I think we'd want to return a `None` here?

**Review by avantgardnerio — APPROVED** [review]

File: datafusion/core/src/physical_optimizer/pruning.rs [hunk 2]
```diff
@@ -721,23 +730,20 @@ fn build_predicate_expression(
     let (left, op, right) = match expr {
         Expr::BinaryExpr(BinaryExpr { left, op, right }) => (left, *op, right),
         Expr::IsNull(expr) => {
-            let expr = build_is_null_column_expr(expr, schema, required_columns)
+            return build_is_null_column_expr(expr, schema, required_columns)
                 .unwrap_or(unhandled);
-            return Ok(expr);
```
[the remark this answers is not in the document]
Comment (crepererum) [reply datafusion/core/src/physical_optimizer/pruning.rs:726]: Yes

File: datafusion/core/src/physical_plan/file_format/parquet/page_filter.rs (hunk 1 above)
Comment (crepererum) [reply datafusion/core/src/physical_plan/file_format/parquet/page_filter.rs:117]: Yeah, we skip the predicate if we don't refer to any column. However you might be right (at least this is how I read your comment) that we need additional test coverage for a "constant" predicate (i.e. one that doesn't reference any column). I'll check next week if such a test exists and if not, add one.

**Pull request merged**